### PR TITLE
CHAT-416 Moving from one space to another empty space, all the discus…

### PR DESCRIPTION
**Pull request by boubaker**
…sion is copied

### CODE REVIEW

File: application/src/main/webapp/js/chat-modules.js [hunk 1]
```diff
@@ -271,16 +271,18 @@ ChatRoom.prototype.refreshChat = function(forceRefresh, callback) {
...
+      if(!dataMessages) {
+        dataMessages = [];
       }
+      var updatedTS = Math.max.apply(Math,TAFFY(dataMessages)().select("lastUpdatedTimestamp").filter(Boolean));
+      if (updatedTS < 0) updatedTS = 0;
+      jzStoreParam("lastTS"+thiss.username, ts, 600);
+      jzStoreParam("lastUpdatedTS"+thiss.username, updatedTS, 600);
+
+      thiss.addMessagesToLocalList(data);
+      thiss.showMessages();
```
Comment (thomasdelhomenie) [inline application/src/main/webapp/js/chat-modules.js]: The showMessages is executed on each refresh, so it means that even if there is no new message, the HTML of the room is re-generated every X seconds. It is not good for performances, and it also adds a blinking effect on the room, especially on the avatars, which is very annoying.

**Comment by thomasdelhomenie** [comment]
It is far better. But still some remarks :
* I think we should display a loading image when switching from one room to another instead of simply empty the messages, it would avoid displaying the "No message" text (which is very visible with bad network). It does not make sense to have this message. WDYT ?
* emptying the room messages should be done earlier IMO. With this PR we still have some time between the click on a room and the emptying of the room. I think we can do better.
* there is an issue when creating a room. After the creation, the new room is selected (which is fine), but the messages are still the messages of the previous room.
* we still have the issue of sending a message to the wrong room when switching between room. It is easily reproducible by emulating a very bad network (with Chrome Dev Tools for example) : create 2 rooms, select the first room, type a message (without sending it), select the second room and send the message immediately (it should be done before of the load of the messages of the selected room) -> the message is sent to first room. I think it is closely related to the current issue, so it is a good opportunity to fix it.

**Comment by boubaker** [comment]
@thomasdelhomenie , those issues are already in place before the patch and this fix will just avoid to have the messages of previous room when moving to an empty room. IMO, we should open a different issue for each of your remark different from this one. WDYT ?

**Comment by thomasdelhomenie** [comment]
Leaving the fix as is feels like an unfinished work IMO.
Point 1, 2 and 3 are clearly part of this issue from my point of view. Maybe we can move the point 4 to a dedicated issue, but it is really related to this issue, so I think it is a good opportunity to fix it.
@hghorbel @plamarque any opinion ?

**Comment by boubaker** [comment]
Point 1 & 2 are functional improvements
Point 3 & 4 are reproduced even without this PR

**Comment by thomasdelhomenie** [comment]
> Point 1 & 2 are functional improvements

I don't agree. We don't have point  currently, the "No message" is not displayed when switching rooms, so it is a regression.
Point 2 is about making this fix better.

> Point 3 & 4 are reproduced even without this PR

For point 3, this is the purpose of this issue to remove message when switching rooms, so if we do not empty messages when creating a new room, the issue is still there and the fix is not complete IMO.
For point 4, as I already said, we can consider it as a new bug, but I think we should use this opportunity to fix it.

**Comment by boubaker** [comment]
> I don't agree. We don't have point currently, the "No message" is not displayed when switching rooms, so it is a regression.

Yep, we had even worst, before. The messages of the old room are still displayed in the new selected room until the new room messages are loaded. So I don't think it's a regression, but an improved and more coherent behavior.

> For point 3, this is the purpose of this issue to remove message when switching rooms

ok, I will consider this.

> For point 4, as I already said, we can consider it as a new bug, but I think we should use this opportunity to fix it.

IMO, it should be treated in an issue apart, because it's not the purpose of the current issue. But I will fix it anyway, no problem ;-)

**Comment by thomasdelhomenie** [comment]
> Yep, we had even worst, before. The messages of the old room are still displayed in the new selected room until the new room messages are loaded. So I don't think it's a regression, but an improved and more coherent behavior.

So it means that it some cases (especially with bad network), the user will see the "No message" during few seconds whereas the room does have messages. It looks very ugly to me, as if the fix is not completely done.

**Comment by boubaker** [comment]
ok, I have seen with @hghorbel and improved the overall user experience in Chat including in very bad network bandwidth.

File: application/src/main/webapp/js/chat-modules.js [hunk 2]
```diff
@@ -197,28 +197,45 @@ ChatRoom.prototype.emptyChatZone = function() {
   jqchat("#chat-record-button .btn").attr("disabled","");
   var $chats = jqchat("#chats");
   $chats.empty();
-  $chats.append("<div class=\"noContent\"><span class=\"text\">" + chatBundleData["exoplatform.chat.no.conversation"] + "</span></div>");
+
+  // Show loading image if we are loading the messages of the room, else display "No messages"
+  if(showLoading) {
+    $chats.append("<div style='text-align:center'><span class='text'><img src='/chat/img/sync.gif' width='64px' class='chatSync' style='margin-top: 176px;'></span></div>");
```
Comment (thomasdelhomenie) [inline application/src/main/webapp/js/chat-modules.js]: Style should be moved in css file.

File: application/src/main/webapp/js/chat.js [hunk 3]
```diff
@@ -1218,6 +1218,8 @@ var handleRoomNotifLayout = function() {
         var teamName = data.name;
         var roomId = "team-"+data.room;
         chatApplication.refreshWhoIsOnline(roomId, teamName);
+        // Refresh the chatRoom after creating the team room
+        chatApplication.chatRoom.refreshChat(true);
```
Comment (thomasdelhomenie) [inline application/src/main/webapp/js/chat.js]: Shouldn't we call loadRoom instead to make sure the room is well loaded ?

File: application/src/main/webapp/js/chat-modules.js [hunk 4]
```diff
@@ -197,28 +197,45 @@ ChatRoom.prototype.emptyChatZone = function() {
   jqchat("#chat-record-button .btn").attr("disabled","");
   var $chats = jqchat("#chats");
   $chats.empty();
-  $chats.append("<div class=\"noContent\"><span class=\"text\">" + chatBundleData["exoplatform.chat.no.conversation"] + "</span></div>");
+
+  // Show loading image if we are loading the messages of the room, else display "No messages"
+  if(showLoading) {
+    $chats.append("<div class='wrapper'><span class='text'><img src='/chat/img/sync.gif' width='64px' class='chatSync' style='margin-top: 176px;'></span></div>");
```
Comment (thomasdelhomenie) [inline application/src/main/webapp/js/chat-modules.js]: Still some style on the image. It could be moved to css file in chatSync class.

Comment (boubaker) [inline application/src/main/webapp/js/chat-modules.js]: It's not a style, it' an html dom attribute. This class is used in another location too, so no need to move it this html attribute to a css class IMO.

File: application/src/main/webapp/js/chat.js (hunk 3 above)
Comment (boubaker) [inline application/src/main/webapp/js/chat.js]: fixed

File: application/src/main/webapp/js/chat-modules.js (hunk 4 above)
Comment (boubaker) [inline application/src/main/webapp/js/chat-modules.js]: fixed

**Comment by boubaker** [comment]
> there is a weird blinking effect on the loading image (you can see it easily with bad network settings). Maybe the DOM is refreshed for nothing ?

This comes from the image itself, you can test:
http://localhost:8080/chat/img/sync.gif

**Comment by thomasdelhomenie** [comment]
> This comes from the image itself, you can test:
http://localhost:8080/chat/img/sync.gif

Indeed :scream_cat: . Maybe Mohamed can do something about it ?

It works fine now with 1:1 rooms :+1: but I still have others issues :
* when I select a room, the loading image appears, then once the messages are retrieved they are displayed, but the message composer is still disabled for few seconds
* when I select a room, once the messages are loaded, the room title disappears

**Comment by boubaker** [comment]
Merged to develop

**Pull request closed without merging**